File: encode.py
import os
import subprocess
import sys

import cv2
import numpy as np
# hhh

# 编码二维码

def read_bin_file(file_name):
    with open(file_name, 'rb') as f:
        data = f.read()
        # print(data)
        # print(str(data))
    return data

def draw_qr_code(data: str, img: list[np.ndarray]):
    key = 0
    while True:
        src = np.full((780, 780, 3), (255, 255, 255), dtype=np.uint8)

        cv2.rectangle(src, (0, 0), (69, 69), (0, 0, 0), -1)
        cv2.rectangle(src, (10, 10), (59, 59), (255, 255, 255), -1)
        cv2.rectangle(src, (20, 20), (49, 49), (0, 0, 0), -1)

        cv2.rectangle(src, (710, 0), (779, 69), (0, 0, 0), -1)
        cv2.rectangle(src, (720, 10), (769, 59), (255, 255, 255), -1)
        cv2.rectangle(src, (730, 20), (759, 49), (0, 0, 0), -1)

        cv2.rectangle(src, (0, 710), (69, 779), (0, 0, 0), -1)
        cv2.rectangle(src, (10, 720), (59, 769), (255, 255, 255), -1)
        cv2.rectangle(src, (20, 730), (49, 759), (0, 0, 0), -1)

        cv2.rectangle(src, (690, 690), (739, 739), (0, 0, 0), -1)
        cv2.rectangle(src, (700, 700), (729, 729), (255, 255, 255), -1)
        cv2.rectangle(src, (710, 710), (719, 719), (0, 0, 0), -1)

        bits = 8
        number = len(img)
        for col in range(78):
            for row in range(78):
                if key >= len(data):
                    background = np.full((860, 860, 3), (255, 255, 255), dtype=np.uint8)
                    background[40:820, 40:820] = src
                    img.append(background)
                    return
                if col < 8 and (row < 8 or row > 69) or (col > 69 and row < 8) or (
                        67 < col < 75 and 67< row < 75):
                    continue

                else:
                    if data[key] == '1':
                        cv2.rectangle(src, (col * 10, row * 10), (col * 10 + 9, row * 10 + 9), (0, 0, 0), -1)
                    else:
                        cv2.rectangle(src, (col * 10, row * 10), (col * 10 + 9, row * 10 + 9), (255, 255, 255), -1)
                    key += 1
        background = np.full((860, 860, 3), (255, 255, 255), dtype=np.uint8)
        background[40:820, 40:820] = src
        img.append(background)



def pic_to_video(img,frame_rate,output):
    for i, frame in enumerate(img):
        cv2.imwrite(f"./assets/img/frame_{i:04d}.png", frame)
    subprocess.run(
        ["ffmpeg", "-framerate", f"{frame_rate}", "-i", f"./assets/img/frame_%04d.png", "-c:v", "libx264", "-pix_fmt",
         "yuv420p", f"./assets/video/{output}"])
    for i in range(len(img)):
        os.remove(f"./assets/img/frame_{i:04d}.png")


def main():
    filename,output,frame_rate=sys.argv[1],sys.argv[2],sys.argv[3]
    string = read_bin_file(f"./assets/bin/{filename}")

    img = []

    draw_qr_code(str(string), img)
    print(len(img))

    pic_to_video(img,frame_rate,output)
    # 保存视频


if __name__ == "__main__":
    main()

File: src/encode.cpp
#include "encode.h"

void ReadBinFile(string& str, const string fileName) {
	ifstream inFile;
	inFile.open(fileName, ios::binary);//打开文件

	if (!inFile) {
		cout << "Unable to open file!" << endl;
		exit(0);
	}

	//读入文件内容
	stringstream sStream;
	sStream << inFile.rdbuf();
	str = sStream.str();
	//cout << str << endl;
}

int en = 0;

void DrawQRCode(string data1, vector<Mat>& img, const string limitTime) {
	int frame_rate = 20;
	double target_duration_seconds = stoi(limitTime) * 1.0 / 1000;
	// 计算最大帧数（向上取整）  
	int target_frames = static_cast<int>(std::ceil(target_duration_seconds * frame_rate));
	int key = 0;
	int num = 0;
	vector<uchar>data;
	for (int i = 0; i < data1.size(); i++)
		data.push_back(data1[i]);
	while (true) {
		Mat src(1440, 1440, CV_8UC3, Scalar(255, 255, 255));

		//左上定位点
		rectangle(src, Point(0, 0), Point(69, 69), Scalar(0, 0, 0), -1);
		rectangle(src, Point(10, 10), Point(59, 59), Scalar(255, 255, 255), -1);
		rectangle(src, Point(20, 20), Point(49, 49), Scalar(0, 0, 0), -1);
		//右上定位点
		rectangle(src, Point(1370, 0), Point(1439, 69), Scalar(0, 0, 0), -1);
		rectangle(src, Point(1380, 10), Point(1429, 59), Scalar(255, 255, 255), -1);
		rectangle(src, Point(1390, 20), Point(1419, 49), Scalar(0, 0, 0), -1);
		//左下定位点
		rectangle(src, Point(0, 1370), Point(69, 1439), Scalar(0, 0, 0), -1);
		rectangle(src, Point(10, 1380), Point(59, 1429), Scalar(255, 255, 255), -1);
		rectangle(src, Point(20, 1390), Point(49, 1419), Scalar(0, 0, 0), -1);

		//右下辅助点
		rectangle(src, Point(1350, 1350), Point(1399, 1399), Scalar(0, 0, 0), -1);
		rectangle(src, Point(1360, 1360), Point(1389, 1389), Scalar(255, 255, 255), -1);
		rectangle(src, Point(1370, 1370), Point(1379, 1379), Scalar(0, 0, 0), -1);

		int bits = 8;
		int flag = 0;
		int page = 0;
		int real_page = 0;
		//绘制数据信息不使用复杂的方法,直接按行列的形式填充
		while (true) {
			real_page++;
			page = real_page;
			for (int col = 0; col < 144; col++) {
				for (int row = 0; row < 144; row++) {
					if (data.size() == key)
					{
						flag++;
						data.push_back('\0');
						data.push_back('\0');
					}
					if (flag == 2) {
						Mat backGround(1600, 1600, CV_8UC3, Scalar(255, 255, 255));
						Rect area(40, 40, 1440, 1440);
						src.copyTo(backGround(area));
						//imwrite(to_string(en++) + ".png", backGround);
						num++;
						if (num > target_frames)
						{
							return;
						}
						img.push_back(backGround);
						return;
					}
					if ((col < 8 && (row < 8 || row > 135)) || (col > 135 && row < 8) || (col > 123 && col < 141 && row>123 && row < 141)) {
						continue;
					}
					else if (col == 143 && row > 56)
						continue;
					else if (col == 0 && row < 16) {
						if (page % 2 == 1) {
							rectangle(src, Point(col * 10, row * 10), Point(col * 10 + 9, row * 10 + 9), Scalar(0, 0, 0), -1);
						}
						else {
							rectangle(src, Point(col * 10, row * 10), Point(col * 10 + 9, row * 10 + 9), Scalar(255, 255, 255), -1);
						}
						page /= 2;
					}
					else {
						if (data.at(key) % 2 == 1) {
							rectangle(src, Point(col * 10, row * 10), Point(col * 10 + 9, row * 10 + 9), Scalar(0, 0, 0), -1);
						}
						else {
							rectangle(src, Point(col * 10, row * 10), Point(col * 10 + 9, row * 10 + 9), Scalar(255, 255, 255), -1);
						}
						bits--;
						data.at(key) /= 2;
					}
					if (bits == 0)
					{
						bits = 8;
						key++;
					}
				}
			}
			//imwrite("1.png", src);
			Mat backGround(1600, 1600, CV_8UC3, Scalar(255, 255, 255));
			Rect area(80, 80, 1440, 1440);
			src.copyTo(backGround(area));
			//imwrite(to_string(en++) + ".png", backGround);
			num++;
			if (num > target_frames)
			{
				return;
			}
			img.push_back(backGround);
		}
	}

	//cout << key << endl;

	//namedWindow("IMAGE");
	//imshow("IMAGE",src);
	//waitKey(0);
}

void PicToVideo(vector<Mat>& img, const string fileName) {
	int frame_rate = 20;
	VideoWriter video(fileName, VideoWriter::fourcc('m', 'p', '4', 'v'), frame_rate, Size(1600, 1600));

	for (int i = 0; i < img.size(); i++) {
		Mat image = img[i].clone();
		video << image;
	}
}
